### ISSUE TITLE
How to use qsim with TFQ?

### ISSUE THREAD

**Issue by perak** [tensorflow/quantum#221]
Hi,

We are running simple benchmark on multiple simulators and both Cirq and TFQ gave us almost exactly the same performance results, so we assumed TFQ is using Cirq's simulator.

Question is: how to turn on qsim to run with TFQ, instead Cirq's simulator?

We are using CPU only machine (no GPU's/TPU's) if that matters.

Source code of QFT benchmark is here:

```python
# Common
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import time
import gc

# Qiskit
from qiskit import QuantumRegister, ClassicalRegister
from qiskit import QuantumCircuit, execute, Aer
from quantastica.qiskit_toaster import ToasterBackend

# pyQuil
from pyquil import Program, get_qc
from pyquil.gates import H, CPHASE, SWAP, MEASURE

# Cirq and TFQ
#import tensorflow_quantum as tfq
import cirq


def qft_qiskit(n):
    qc = QuantumCircuit()

    q = QuantumRegister(n, 'q')
    c = ClassicalRegister(n, 'c')

    qc.add_register(q)
    qc.add_register(c)

    for j in range(n):
        for k in range(j):
            qc.cu1(math.pi/float(2**(j-k)), q[j], q[k])
        qc.h(q[j])

    for i in range(n):
        qc.measure(q[i], c[i])
        
    return qc


def qft_pyquil(n):
    p = Program()

    ro = p.declare('ro', memory_type='BIT', memory_size=n)

    for j in range(n):
        for k in range(j):
             p.inst(CPHASE(math.pi/float(2**(j-k)), j, k))
        p.inst(H(j))

    for i in range(n):
        p.inst(MEASURE(i, ro[i]))
        
    return p


def qft_cirq(n):
    q = cirq.GridQubit.rect(1, n)

    gates = []
    for j in range(n):
        for k in range(j):
             gates.append(cirq.ZPowGate(exponent=math.pi/float(2**(j-k))).controlled()(q[j], q[k]))
        gates.append(cirq.H(q[j]))

    for i in range(n):
        gates.append(cirq.measure(q[i], key='c' + str(i)))
    
    circ = cirq.Circuit(gates)

    return circ


def benchmark_qft_qiskit(from_qubits, to_qubits, results):
    gc.disable()

    results["Qiskit-Aer"] = np.nan
    results["Qiskit-Toaster"] = np.nan

    aer_backend = Aer.get_backend('qasm_simulator')
    toaster_backend = ToasterBackend.get_backend('qasm_simulator')

    for i in range(from_qubits, to_qubits + 1):
        circ = qft_qiskit(i)

        # Repeat multiple times for small number of qubits and get average time
        repeat = 4 if i <= 20 else 1

        # Qiskit with Toaster backend
        toaster_time = np.nan
        for r in range(repeat):
            gc.collect()
            start_time = time.time()
            job = execute(circ, backend=toaster_backend, shots=1)
            result = job.result()
            elapsed_time = (time.time() - start_time) * 1000
            toaster_time = elapsed_time if np.isnan(toaster_time) else min(toaster_time, elapsed_time)
        results["Qiskit-Toaster"][i] = toaster_time

        # Qiskit with Aer backend
        aer_time = np.nan
        for r in range(repeat):
            gc.collect()
            start_time = time.time()
            job = execute(circ, backend=aer_backend, shots=1)
            result = job.result()
            elapsed_time = (time.time() - start_time) * 1000
            aer_time = elapsed_time if np.isnan(aer_time) else min(aer_time, elapsed_time)
        results["Qiskit-Aer"][i] = aer_time

    gc.enable()


def benchmark_qft_pyquil(from_qubits, to_qubits, results):
    gc.disable()

    results["pyQuil-QVM"] = np.nan

    for i in range(from_qubits, to_qubits + 1):

        circ = qft_pyquil(i)
        circ.wrap_in_numshots_loop(1)

        # Repeat multiple times for small number of qubits and get average time
        repeat = 4 if i <= 20 else 1

        # pyQuil with QVM backend
        qc = get_qc(str(i) + 'q-qvm')
        
        qvm_time = np.nan
        for r in range(repeat):
            gc.collect()
            start_time = time.time()
            result = qc.run(circ)        
            elapsed_time = (time.time() - start_time) * 1000
            qvm_time = elapsed_time if np.isnan(qvm_time) else min(qvm_time, elapsed_time)
        results["pyQuil-QVM"][i] = qvm_time

    gc.enable()

        
def benchmark_qft_cirq(from_qubits, to_qubits, results):
    gc.disable()

    results["Cirq-Simulator"] = np.nan
    #results["TFQ"] = np.nan

    for i in range(from_qubits, to_qubits + 1):
        simulator = cirq.Simulator()

        circ = qft_cirq(i)

        # Repeat multiple times for small number of qubits and get average time
        repeat = 4 if i <= 20 else 1

        # Cirq simulator
        cirq_time = np.nan
        for r in range(repeat):
            gc.collect()
            start_time = time.time()
            result = simulator.run(circ, repetitions=1)
            elapsed_time = (time.time() - start_time) * 1000
            cirq_time = elapsed_time if np.isnan(cirq_time) else min(cirq_time, elapsed_time)
        results["Cirq-Simulator"][i] = cirq_time

        """
        # TFQ simulator
        tfq_time = np.nan
        for r in range(repeat):
            gc.collect()
            start_time = time.time()
            result = tfq.layers.Sample()(circ, repetitions=1)
            elapsed_time = (time.time() - start_time) * 1000
            tfq_time = elapsed_time if np.isnan(tfq_time) else min(tfq_time, elapsed_time)
        results["TFQ"][i] = tfq_time
        """
    gc.enable()


def benchmark_qft(from_qubits, to_qubits):
    results = pd.DataFrame(index=range(from_qubits, to_qubits + 1), columns=[])
    
    # Qiskit
    benchmark_qft_qiskit(from_qubits, to_qubits, results)
    
    # pyQuil
    benchmark_qft_pyquil(from_qubits, to_qubits, results)

    # Cirq
    benchmark_qft_cirq(from_qubits, to_qubits, results)

    print(results)

    plt.figure(dpi=300)
    plt.title("QFT")
    plt.xlabel("Qubits")
    plt.ylabel("Time (ms)")
    plt.xticks(range(from_qubits, to_qubits + 1, 2))
    plt.yscale("log")
    plt.plot(results)
    plt.grid(which='both')
    plt.legend(loc="upper left", labels=results.columns)
    plt.savefig("output/benchmark_qft.png")


benchmark_qft(1, 27)
```

And results looks something like this:

![benchmark_qft](https://user-images.githubusercontent.com/5630814/80431745-8f8bfa00-88f2-11ea-8e29-faeecdb295e3.png)

TFQ is temporary removed from benchmark as it shows exactly the same performance. We wish to turn on TFQ's internal simulator (qsim?) ans see how it performs compared to others.

Thanks!

**Comment by perak** [comment]
I forgot to show versions:

```
cirq==0.7.0
tensorflow==2.1.0
tensorflow-estimator==2.1.0
tensorflow-quantum==0.2.0
qiskit==0.18.0
qiskit-aer==0.5.0
qiskit-aqua==0.6.5
qiskit-ibmq-provider==0.6.0
qiskit-ignis==0.3.0
qiskit-terra==0.13.0
quantastica-qiskit-toaster==0.9.24
pyquil==2.19.0
```

**Comment by MichaelBroughton** [comment]
qsim is enabled by default what you are running was using qsim, but there are a couple of things going on here:

1. You are using layers which have a bunch of convenience code in place that do make things much slower. For example inside of you're timing benchmark you are converting a `cirq.Circuit` into a `tf.Tensor` everytime which isn't necessary. This goes against the common workflow in TF and TFQ of calling `tf._convert_to_tensor` or `tfq.convert_to_tensor` once then simulating and operating on the tensor many times. To improve this I would recommend you call `convert_to_tensor` on your input circuit before it gets timed.

2. TFQ Wasn't really designed for single circuit simulation speed. It was designed for many circuit simulation speed. We didn't do a lot of single circuit testing so it's likely not super optimized in that regard.

3. You are most likely operating in eager mode (slower than graph mode).

4. Another possible thing that is hurting TFQ is that you are changing the qubit numbers quite drastically which will cause `tf.function` retracing to occur also slowing things down.

5. In the past I remember looking into how fast our parsing code was and I remember it not being all that great.

My hunch is that 2. 4. and 5. are hurting TFQ the most. Could you change the benchmark circuit to something with the same number of qubits, but far fewer total gates and see what happens ? That would tell us if it's #5 or not.

**Comment by perak** [comment]
@MichaelBroughton thank you for prompt answer.

Problem was that TFQ and Cirq show (almost) exactly the same performance with any number of qubits, so it look like results are coming from the same simulator. This is what we have when running (the same benchmark code) on Google Cloud `n1-standard-8` :

![benchmark_qft (2)](https://user-images.githubusercontent.com/5630814/80570777-acf8bb00-89fb-11ea-8b68-411a2a06dc5e.png)

My guess is that slight difference comes from different handling/processing of results, not from simulation algorithm itself. 

So, I still believe that the two use the same simulation algorithm (or the same simulator).

However, as this is my first contact with TFQ, must be I am doing something wrong. 

Would you be kind to show me how to properly run simulation "TFQ way"?

```python
circ = cirq.Circuit // ...

result = tfq.layers.Sample()(circ, repetitions=1)
```

Goal of our benchmark is to compare simulator algorithms, so we wish to simply run simulation with 1...N qubits on different simulators and measure time.

Thanks!

**Comment by MichaelBroughton** [comment]
OK. I took the code you pasted and made some quick changes:
```python
import tensorflow_quantum as tfq
import cirq
import time
import math
import matplotlib.pyplot as plt
import numpy as np

def qft_cirq(n):
    q = cirq.GridQubit.rect(1, n)

    gates = []
    for j in range(n):
        for k in range(j):
             gates.append(cirq.ZPowGate(exponent=math.pi/float(2**(j-k))).controlled()(q[j], q[k]))
        gates.append(cirq.H(q[j]))

    circ = cirq.Circuit(gates)
    return circ

ct = []
tfqt = []
for k in range(5, 23):
    reference_circuit = cirq.generate_boixo_2018_supremacy_circuits_v2(cirq.GridQubit.rect(1, k), 25, seed=1234)
    # reference_circuit = qft_cirq(k)
    pre_converted = tfq.convert_to_tensor([reference_circuit])
    t = time.time()
    s = tfq.layers.State()(pre_converted)
    tfqt.append(time.time() - t)

    t = time.time()
    s = cirq.Simulator().simulate(reference_circuit)
    ct.append(time.time() - t)

plt.plot(np.log(ct), label='Cirq')
plt.plot(np.log(tfqt), label='TFQ')
plt.legend()
plt.show()
```
I've opted to simulate the state since simulating the state and then sampling is testing more moving parts than necessary. I also switched to seconds since ms seems like overkill.

TFQ was designed to be fast for BATCHES of circuits not a SINGLE circuit. So when we run our simulator in your example, here is roughly what is happening in TFQ:
1. The `tfq_simulate_state` op picks up the serialized circuit.
2. The circuit gets deserialized to a proto object.
3. We convert the proto object to a series of matrices.
4. We see if we can run gate fusion over these matrices ( since it's QFT we don't find much to fuse)
5. We loop over the circuit gate by gate and multiply it onto the state.
6. We Check for padding and write the output state to the final tensor.

This is done for each circuit in a batch of circuits that gets fed into TFQ, if the batch is of size 1 we wouldn't expect much boost.

in Cirq the state is represented by a numpy array and to simulate a circuit the following things are done:
1. The circuit is looped over gate by gate and if an apply_unitary method exists that is called otherwise the raw matrix is multiplied onto the state.
2. The final state is returned.

We see the speed is around 1:1 since TFQ has all these extra steps, fusing doesn't cut down the gate count and since numpy array operations break down into fast C code. This seems fine. In this simulation and yours qsim IS being used.

![image](https://user-images.githubusercontent.com/2791105/80582534-be25d580-89c3-11ea-8f4a-0473af8fd051.png)

Now lets modify your code to use multiple values in placeholders in a circuit and see how TFQ does:

```python
import tensorflow_quantum as tfq
import cirq
import sympy
import time
import math
import matplotlib.pyplot as plt
import numpy as np

def qft_cirq(n):
    q = cirq.GridQubit.rect(1, n)
    alpha = sympy.Symbol('alpha')

    gates = []
    for j in range(n):
        for k in range(j):
             gates.append(cirq.ZPowGate(exponent=alpha).controlled()(q[j], q[k]))
        gates.append(cirq.H(q[j]))

    circ = cirq.Circuit(gates)
    return circ

ct = []
tfqt = []
for k in range(5, 23):
    # reference_circuit = cirq.generate_boixo_2018_supremacy_circuits_v2(cirq.GridQubit.rect(1, k), 25, seed=1234)
    reference_circuit = qft_cirq(k)
    t = time.time()
    s = tfq.layers.State()(reference_circuit, symbol_names=['alpha'], symbol_values = [[float(i)] for i in range(10)])
    tfqt.append(time.time() - t)

    t = time.time()
    s = cirq.Simulator().simulate_sweep(reference_circuit, params=[{'alpha': float(i)} for i in range(10)])
    ct.append(time.time() - t)

plt.plot(np.log(ct), label='Cirq')
plt.plot(np.log(tfqt), label='TFQ')
plt.legend()
plt.show()
```

![image](https://user-images.githubusercontent.com/2791105/80584056-00501680-89c6-11ea-918d-97a1b0c61f4f.png)

Here's a nice 7-10x speed boost on simulating your problem with multiple different parameter values requiring BATCH simulation (though it isn't exactly a qft anymore, but that doesn't affect simulation time here). 

Feel free to try these with your qft circuits and the supremacy style circuits. You will see a drastic improvement on the supremacy style circuits since we can fuse a few more gates. This is where TFQ shines. **Batches of circuits**.

Based on your issue title, if what you are after is the best performing version of qsim for a SINGLE circuit you might want to `pip install qsimcirq` and then do this:

```python
import qsimcirq
import cirq
import time
import math
import matplotlib.pyplot as plt
import numpy as np

def not_exactly_qft_cirq(n):
    q = cirq.GridQubit.rect(1, n)

    gates = []
    for j in range(n):
        for k in range(j):
             gates.append(cirq.CZ(q[j], q[k]))
        gates.append(cirq.H(q[j]))

    circ = cirq.Circuit(cirq.decompose(gates))
    return circ

ct = []
qst = []
for k in range(5, 23):
    #reference_circuit = cirq.generate_boixo_2018_supremacy_circuits_v2(cirq.GridQubit.rect(1, k), 25, seed=1234)
    reference_circuit = not_exactly_qft_cirq(k)
    t = time.time()
    qsimcirq.QSimSimulator().simulate(qsimcirq.QSimCircuit(reference_circuit))
    qst.append(time.time() - t)

    t = time.time()
    s = cirq.Simulator().simulate(reference_circuit)
    ct.append(time.time() - t)

plt.plot(np.log(ct), label='Cirq')
plt.plot(np.log(qst), label='QSIM')
plt.legend()
plt.show()
```

![image](https://user-images.githubusercontent.com/2791105/80585758-bb79af00-89c8-11ea-8428-3b4869b5e037.png)

Which will give a 7-10x boost for a single circuit.

So in summary: TFQ is at it's fastest on BATCHES of circuits and regular qsim is at it's fastest for SINGLE circuits. If you want an apples to apples comparison of simulation for SINGLE circuits I would recommend you use `qsimcirq` if you want to compare batch simulation speed I would recommend you use TFQ. The behavior you are running into in your benchmarks makes sense because your benchmarks are for SINGLE circuits and `qsimcirq` is better than TFQ for single circuits.

Does this clear things up ?

**Comment by perak** [comment]
@MichaelBroughton thank you very much for detailed clarification! <3

Benchmark repository is now public and  located here: https://github.com/quantastica/quantum-simulator-benchmark

So you can track results if you are interested.

We will add `qsimcirq` and hopefully more simulators (and more circuits, not only QFT).

I'm closing this issue now, thank you!

**Comment by perak** [comment]
P.S. looks like there is no translation from Cirq to qsim for frequently used gates like Measure (!?) and CPhase.

We'll try to use qsim native input format instead converting from cirq, but that requires us to focus and spend time on this, so we'll skip adding both TFQ and qsim to benchmark for now.

**Comment by perak** [comment]
OK, added qsim to benchmark, conversation is moved here: https://github.com/quantumlib/qsim/issues/88
